src/intptrcast.rs

```diff
@@ -63,6 +65,11 @@ impl GlobalStateInner {
 }
 
 impl<'mir, 'tcx> GlobalStateInner {
+    pub fn remove_unreachable_allocs(&mut self, live_allocs: &FxHashSet<AllocId>) {
+        self.int_to_ptr_map.retain(|(_, id)| live_allocs.contains(id));
```

Comment:
Removing things from `int_to_ptr_map` will mean we return a different `AllocId` when an address inside that allocation is looked up. Why does that not change the semantics?

Reply:
I've been unable to craft an example that produces any problematic behavior. I think this data structure is only used when we are immediately using an absolute pointer, either to deref it or to reallocate/deallocate it. So we can fail to select an allocation instead of finding the dead allocation, which doesn't change whether the interpreter step is about to report UB.

But I've spent so much time thinking about this situation, this code is so complicated in how everything cooperates indirectly. I might see if there's a simpler way to write this that's more obviously consistent. I think we're going to have all these same problems when we try to reuse addresses.

Reply:
>  I think this data structure is only used when we are immediately using an absolute pointer, either to deref it or to reallocate/deallocate it. So we can fail to select an allocation instead of finding the dead allocation, which doesn't change whether the interpreter step is about to report UB.

Yes I think that's true. But does "unreachable" imply "dead"? You're removing AllocId that aren't mentioned anywhere, but what if an actually existing allocation isn't mentioned anywhere?

Also, can it change which kind of UB is reported? That wouldn't be great either.

I think for now it'd be better to now remove things from `int_to_ptr_map` ever. Which probably means we should add the `AllocId` in that map to the live set.

> But I've spent so much time thinking about this situation, this code is so complicated in how everything cooperates indirectly. I might see if there's a simpler way to write this that's more obviously consistent. I think we're going to have all these same problems when we try to reuse addresses.

I did spend a lot of time on our integer address  system to make it as nice as I could, but if you have ideas for cleaning it up further then I'm curious to see them. :) The tricky part is having all the generic code in the interpreter also work for pointers with relative addresses (for const-eval).

Reply:
> But does "unreachable" imply "dead"?

No, but I've fixed this now by adding all live AllocIds to the set. Does that address your other concerns here? `GlobalStateInner::alloc_id_from_addr` checks that the AllocId it is going to return is live, so what we do with dead allocations should be irrelevant.

Reply:
Ah, and `alloc_id_from_addr` also checks that we are in-bounds. So it's always an in-bounds address of a live allocation that we return, and therefore we can indeed remove dead allocations. Yes that should work.

Reply:
Ah no, there's still one case where this changes behavior. If we have two allocations directly adjacent to each other, then the address "at the end" of the first equals the base address of the 2nd. Once the 2nd allocation becomes dead, as long as it remains in the `int_to_ptr_map`, `alloc_id_from_addr` will return `None` for the base address of the 2nd allocation. But if we remove the 2nd allocation from the `int_to_ptr_map`, then it will now return `Some(first_alloc_id)`.

Reply:
We're now managing `into_to_ptr_map` without the GC. Do we still need to add all its entries to the reachable set?

Reply:
The GC should probably entirely ignore `int_to_ptr_map` -- neither scan it, nor clean it up.

That said... do we have all live allocations in the reachable set? Cleaning up allocations that are still live seems a bit risky.^^ If we have all live allocations already in there, then scanning `int_to_ptr_map` anyway can't add anything new.

Reply:
Oh yeah 🤦 that's another thing I was going to fix

Reply:
I added some code to explicitly initialize the collection with all live allocation IDs.